Sources/Sentry/SentryNetworkTracker.m

```diff
@@ -0,0 +1,152 @@
+#import "SentryNetworkTracker.h"
+#import "SentryOptions+Private.h"
+#import "SentryPerformanceTracker.h"
+#import "SentrySDK+Private.h"
+#import "SentrySpan.h"
+#import <objc/runtime.h>
+
+static NSString *const SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID
+    = @"SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID";
+
+@interface
+SentryNetworkTracker ()
+
+@property (nonatomic, strong) SentryPerformanceTracker *tracker;
+
+@end
+
+@implementation SentryNetworkTracker
+
++ (SentryNetworkTracker *)sharedInstance
+{
+    static SentryNetworkTracker *instance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{ instance = [[self alloc] init]; });
+    return instance;
+}
+
+- (instancetype)init
+{
+    if (self = [super init]) {
+        self.tracker = SentryPerformanceTracker.shared;
+    }
+    return self;
+}
+
+- (void)urlSessionTaskResume:(NSURLSessionTask *)sessionTask
+{
+    NSURL *url = [[sessionTask currentRequest] URL];
+
+    if (url == nil || ![self isTaskSupported:sessionTask])
+        return;
+
+    NSURL *apiUrl = [NSURL URLWithString:SentrySDK.options.dsn];
+    if ([url.host isEqualToString:apiUrl.host] && [url.path containsString:apiUrl.path])
+        return;
+
+    NSString *statePath = NSStringFromSelector(@selector(state));
+    [sessionTask addObserver:self
+                  forKeyPath:statePath
+                     options:NSKeyValueObservingOptionNew
+                     context:nil];
+
+    SentrySpanId *spanId =
+        [self.tracker startSpanWithName:[NSString stringWithFormat:@"%@ %@",
+                                                  sessionTask.currentRequest.HTTPMethod, url]
+                              operation:SENTRY_NETWORK_REQUEST_OPERATION];
+
+    objc_setAssociatedObject(sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID, spanId,
+        OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey, id> *)change
+                       context:(void *)context
+{
+    if ([keyPath isEqualToString:NSStringFromSelector(@selector(state))]) {
+        NSURLSessionTask *sessionTask = object;
+        if (sessionTask.state != NSURLSessionTaskStateRunning) {
+            SentrySpanId *spanId
+                = objc_getAssociatedObject(sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID);
+
+            if (spanId != nil) {
+                id<SentrySpan> span = [self.tracker getSpan:spanId];
+                NSInteger responseStatusCode = [self urlResponseStatusCode:sessionTask.response];
+
+                [span setDataValue:[NSNumber numberWithInteger:responseStatusCode]
```

Comment:
`h`: This will set the response status code to `-1` if there is none, which doesn't exist. We must check here if it is `-1` and don't add the status code if it is. Can we please also add a test for this?